### PR TITLE
Helper codes for databases

### PR DESCRIPTION
**Pull request by yingwaili**
This pull request consists of codes for two main functionalities:
1. Generate metadata from databases.
2. Generate xyz file from databases.
It currently supports databases with numpy arrays (.npz extension) and HDF5 format (.h5 or .hdf5 extensions).

Codes were originally written by @JaredKeithAveritt and @bnebgen-LANL. Polished by @yingwaili.

### CODE REVIEW

File: examples/convert_db_to_extxyz.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import os
+import sys
+from pathlib import Path
+
+from hippynn.databases.xyz_database_tools import load_base_database, write_extxyz
+
+def main():
+    if len(sys.argv) < 2:
```
Comment (lubbersnick) [inline examples/convert_db_to_extxyz.py:8]: I would strongly prefer that we use argparse, it is easy to set up and contains a number of quality of life features with very little work. (such as `python script.py --help`)

File: hippynn/databases/xyz_database_tools.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+import numpy as np
+import torch
+from ase import Atoms
+from ase.io import write as ase_write
+
+from hippynn.databases import NPZDatabase
+from hippynn.databases.h5_pyanitools import PyAniFileDB
+
+
+def load_base_database(
```
Comment (lubbersnick) [inline hippynn/databases/xyz_database_tools.py:16]: This looks like a good new feature that maybe we should put in a more central location. It would be highly convenient if we could attempt to auto-detect other kinds as well, such as SNAP or ASE databases, so that code changes were minimized when moving between database formats.

File: examples/metadatabase.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    targets = ['energies', 'forces']
+    db_info = {"inputs": inputs , "targets": targets}
+    energies_key_sel = 'energies'
+    base_database = load_db(
+        db_info,
+        en_name,
+        force_name,
+        seed=101,
+        location=DATA_FILE,
+        n_workers=2
```
Comment (lubbersnick) [inline examples/metadatabase.py:85]: Something funny is happening here where this appears to duplicate the intent of your load_base_database in xyz_database_tools ? Is it possible to use that instead?

File: hippynn/databases/xyz_database_tools.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            targets=targets,
+        )
+    
+    else:
+        raise ValueError(f"Unrecognized dataset file extension: {ext}. Supported file extensions are: .h5, .hdf5, .npz.")
+    
+    return db, energies_key
+
+
+def write_extxyz(
```
Comment (lubbersnick) [inline hippynn/databases/xyz_database_tools.py:62]: This is a nice function, similar to write_h5 and write_npz, I think we should add a method to the base database which allows the user to call this from the database object.  It would also be best if the expected keys were encoded as default arguments. For example, coordinates='coordinates' by default, but user can pass coordinates='positions' to use that db_name instead.

File: hippynn/databases/xyz_database_tools.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    pbc can be:
+      - False (default, non-periodic)
+      - True (periodic in all directions)
+      - tuple(bool, bool, bool) for per-axis periodicity
+    """
+    out_path = Path(str(filename))
+    if out_path.exists():
+        if not overwrite:
+            raise FileExistsError(f"Path exists: {out_path}")
+        out_path.unlink()
```
Comment (lubbersnick) [inline hippynn/databases/xyz_database_tools.py:90]: I would not unlink the existing file until everything else has succeeded.

File: hippynn/databases/metadatabase.py [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+                if not quiet: print("Error saving JSON:", e)
+
+        if self.write_metadata_to_csv:
+            try: self.save_metadata_to_csv()
+            except Exception as e:
+                if not quiet: print("Error saving CSV:", e)
+
+    # ─── Plotting ──────────────────────────────────────────────────────────────
+
+    def plot_distributions(
```
Comment (lubbersnick) [inline hippynn/databases/metadatabase.py:768]: There are some hippynn idoms to use with plotting commands that can be helpful here. The typical arguments are shown:bool=False, saved:Union[bool,str]=False. If saved is a str, use that as a file name for saving. If saved is true, use hippynn.settings.DEFAULT_PLOT_FILETYPE along with a default name for the plot. Adding these arguments enables running systems without windowing.

File: hippynn/databases/metadatabase.py [hunk 7]
```diff
@@ -0,0 +1,848 @@
...
+               "pair_first": first,
+               "pair_second": second,
+               "pair_coord": rel}
+        if periodic and cell is not None:
+            offs = frac[second] - frac[first]
+            offs -= np.round(offs)
+            out["cell_offsets"] = offs @ cell_np
+        return out
+
+    def calculate_min_distance(self, periodic=True):
```
Comment (lubbersnick) [inline hippynn/databases/metadatabase.py:431]: We have a routine for this in hippynn.pretraining.calc_min_distances which takes advantage of the graph system, fully GPU,  batchable, etc. With small modifications it could return info for find_pairs as well. I don't know how relevant this is in the intended use case for metadata, but a feature which is important in the MLIAP calculations is that if the cells become small enough then you can have the same two particles have multiple sets of pairs. (similarly, consider what happens if we send a 1-atom cell into this code--i might be mistaken but it would actually return no pairs, right? Is that a problem?)
